File: windows/tests/BabyMonitor.Core.Tests/MissTest.cs
using System.Text;
using BabyMonitor.Core.Json;
using BabyMonitor.Core.Xiaomi;
using Xunit;

namespace BabyMonitor.Core.Tests;

public class MissTest
{
    [Fact(DisplayName = "PROTO-20 auth payload carries public key, sign, empty uuid and no encryption")]
    public void AuthPayload()
    {
        var json = new JsonObj(Miss.AuthPayload("aabbcc", "SIGNTOKEN"));
        Assert.Equal("aabbcc", json.GetString("public_key"));
        Assert.Equal("SIGNTOKEN", json.GetString("sign"));
        Assert.Equal(string.Empty, json.GetString("uuid"));
        Assert.Equal(0, json.GetInt("support_encrypt"));
    }

    [Fact(DisplayName = "PROTO-21 quality maps hd to 2 for the C100 and 3 for the C200/C300")]
    public void QualityMapping()
    {
        Assert.Equal(
            """{"videoquality":2,"enableaudio":1}""",
            Miss.StartMediaBody("chuangmi.camera.077ac1", "hd", true));
        Assert.Equal("""{"videoquality":3,"enableaudio":1}""", Miss.StartMediaBody(Miss.ModelC200, "hd", true));
        Assert.Equal("""{"videoquality":3,"enableaudio":1}""", Miss.StartMediaBody(Miss.ModelC300, "hd", true));
        Assert.Equal(
            """{"videoquality":1,"enableaudio":0}""",
            Miss.StartMediaBody("chuangmi.camera.077ac1", "sd", false));
        Assert.Equal(
            """{"videoquality":0,"enableaudio":1}""",
            Miss.StartMediaBody("chuangmi.camera.077ac1", "auto", true));
    }

    [Fact(DisplayName = "PROTO-21 inner commands are BE u32 command plus body")]
    public void InnerCommands()
    {
        var payload = Miss.InnerCommand(Miss.CmdVideoStart, "{}");
        Assert.Equal(0x102L, payload.BeU32(0));
        Assert.Equal("{}", Encoding.UTF8.GetString(payload.Slice(4, payload.Length)));

        var stop = Miss.InnerCommand(Miss.CmdVideoStop);
        Assert.Equal(4, stop.Length);
        Assert.Equal(0x103L, stop.BeU32(0));
    }

    [Fact(DisplayName = "PROTO-22 media headers parse little-endian size, codec, seq, flags and a u64 pts")]
    public void MediaHeaders()
    {
        var hdr = new byte[32];
        hdr.PutLeU32(0, 512);
        hdr.PutLeU32(4, Miss.CodecH265);
        hdr.PutLeU32(8, 42);
        hdr.PutLeU32(12, 0b1000);
        hdr.PutLeU32(16, 5_000_000_000L & 0xffffffffL);
        hdr.PutLeU32(20, 5_000_000_000L >> 32); // > u32 to prove a 64-bit pts

        var parsed = Miss.ParseMediaHeader(hdr);
        Assert.Equal(512L, parsed.Size);
        Assert.Equal(Miss.CodecH265, parsed.CodecId);
        Assert.Equal(42L, parsed.Sequence);
        Assert.Equal(0b1000L, parsed.Flags);
        Assert.Equal(5_000_000_000L, parsed.PtsMs);
    }

    [Fact(DisplayName = "PROTO-23 codec ids map to frames and unknown ids are skipped")]
    public void CodecMapping()
    {
        var data = new byte[] { 1, 2, 3 };
        static Miss.MediaHeader Header(long codec, long flags = 0) =>
            new(Size: 3, CodecId: codec, Sequence: 1, Flags: flags, PtsMs: 10);

        var h265 = Assert.IsType<Frame.Video>(Miss.FrameFromPacket(Header(Miss.CodecH265), data));
        Assert.Equal("h265", h265.Codec);
        Assert.Equal(data, h265.Data);

        var h264 = Assert.IsType<Frame.Video>(Miss.FrameFromPacket(Header(Miss.CodecH264), data));
        Assert.Equal("h264", h264.Codec);

        var opus = Assert.IsType<Frame.Audio>(Miss.FrameFromPacket(Header(Miss.CodecOpus), data));
        Assert.Equal("opus", opus.Codec);
        Assert.Equal(48000, opus.SampleRate);

        var pcma = Assert.IsType<Frame.Audio>(Miss.FrameFromPacket(Header(Miss.CodecPcma), data));
        Assert.Equal("pcma", pcma.Codec);

        Assert.Null(Miss.FrameFromPacket(Header(9999), data));
    }

    [Fact(DisplayName = "PROTO-23 pcm sample rate derives from flags bits 3-6")]
    public void PcmSampleRate()
    {
        Assert.Equal(8000, Miss.PcmSampleRate(0));
        Assert.Equal(16000, Miss.PcmSampleRate(0b1000));
        Assert.Equal(16000, Miss.PcmSampleRate(0b1110000));
        Assert.Equal(8000, Miss.PcmSampleRate(0b10000000)); // bit 7 is outside the field
    }

    [Fact(DisplayName = "PROTO-11 cameras are devices whose model contains camera")]
    public void CameraModels()
    {
        Assert.True(Mi.IsCamera("chuangmi.camera.077ac1"));
        Assert.True(Mi.IsCamera("isa.camera.hlc7"));
        Assert.False(Mi.IsCamera("zhimi.airpurifier.ma2"));
        Assert.False(Mi.IsCamera("yeelink.light.lamp4"));
    }

    [Fact(DisplayName = "PROTO-24 night-vision modes map to their wire values and back")]
    public void NightVisionModes()
    {
        // Note ON is 0, not 1.
        Assert.Equal(0, (int)NightVisionMode.On);
        Assert.Equal(1, (int)NightVisionMode.Off);
        Assert.Equal(2, (int)NightVisionMode.Auto);

        Assert.Equal(NightVisionMode.On, Mi.NightVisionFromValue(0));
        Assert.Equal(NightVisionMode.Off, Mi.NightVisionFromValue(1));
        Assert.Equal(NightVisionMode.Auto, Mi.NightVisionFromValue(2));
        Assert.Null(Mi.NightVisionFromValue(7)); // an unknown value is null — not a crash
    }
}

File: core/src/commonTest/kotlin/com/bluzi/babymonitor/xiaomi/MissTest.kt
package com.bluzi.babymonitor.xiaomi

import com.bluzi.babymonitor.json.JSONObject
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue
import kotlin.test.Test

class MissTest {
    @Test
    fun `PROTO-20 auth payload carries public key — sign — empty uuid — no encryption`() {
        val json = JSONObject(Miss.authPayload("aabbcc", "SIGNTOKEN"))
        assertEquals("aabbcc", json.getString("public_key"))
        assertEquals("SIGNTOKEN", json.getString("sign"))
        assertEquals("", json.getString("uuid"))
        assertEquals(0, json.getInt("support_encrypt"))
    }

    @Test
    fun `PROTO-21 quality maps hd to 2 for C100 and 3 for C200-C300`() {
        assertEquals("""{"videoquality":2,"enableaudio":1}""", Miss.startMediaBody("chuangmi.camera.077ac1", "hd", true))
        assertEquals("""{"videoquality":3,"enableaudio":1}""", Miss.startMediaBody(Miss.MODEL_C200, "hd", true))
        assertEquals("""{"videoquality":3,"enableaudio":1}""", Miss.startMediaBody(Miss.MODEL_C300, "hd", true))
        assertEquals("""{"videoquality":1,"enableaudio":0}""", Miss.startMediaBody("chuangmi.camera.077ac1", "sd", false))
        assertEquals("""{"videoquality":0,"enableaudio":1}""", Miss.startMediaBody("chuangmi.camera.077ac1", "auto", true))
    }

    @Test
    fun `PROTO-21 inner commands are BE u32 command plus body`() {
        val payload = Miss.innerCommand(Miss.CMD_VIDEO_START, "{}")
        assertEquals(0x102L, payload.beU32(0))
        assertEquals("{}", payload.copyOfRange(4, payload.size).decodeToString())

        val stop = Miss.innerCommand(Miss.CMD_VIDEO_STOP)
        assertEquals(4, stop.size)
        assertEquals(0x103L, stop.beU32(0))
    }

    @Test
    fun `PROTO-22 media headers parse little-endian size codec seq flags and u64 pts`() {
        val hdr = ByteArray(32)
        hdr.putLeU32(0, 512)
        hdr.putLeU32(4, Miss.CODEC_H265)
        hdr.putLeU32(8, 42)
        hdr.putLeU32(12, 0b1000)
        hdr.putLeU64(16, 5_000_000_000L) // > u32 to prove 64-bit pts
        val parsed = Miss.parseMediaHeader(hdr)
        assertEquals(512L, parsed.size)
        assertEquals(Miss.CODEC_H265, parsed.codecId)
        assertEquals(42L, parsed.sequence)
        assertEquals(0b1000L, parsed.flags)
        assertEquals(5_000_000_000L, parsed.ptsMs)
    }

    @Test
    fun `PROTO-23 codec ids map to frames and unknown ids are skipped`() {
        val data = byteArrayOf(1, 2, 3)
        fun header(codec: Long, flags: Long = 0) =
            Miss.MediaHeader(size = 3, codecId = codec, sequence = 1, flags = flags, ptsMs = 10)

        val h265 = Miss.frameFromPacket(header(Miss.CODEC_H265), data) as Frame.Video
        assertEquals("h265", h265.codec)
        assertContentEquals(data, h265.data)

        val h264 = Miss.frameFromPacket(header(Miss.CODEC_H264), data) as Frame.Video
        assertEquals("h264", h264.codec)

        val opus = Miss.frameFromPacket(header(Miss.CODEC_OPUS), data) as Frame.Audio
        assertEquals("opus", opus.codec)
        assertEquals(48000, opus.sampleRate)

        val pcma = Miss.frameFromPacket(header(Miss.CODEC_PCMA), data) as Frame.Audio
        assertEquals("pcma", pcma.codec)

        assertNull(Miss.frameFromPacket(header(9999), data))
    }

    @Test
    fun `PROTO-23 pcm sample rate derives from flags bits 3-6`() {
        assertEquals(8000, Miss.pcmSampleRate(0))
        assertEquals(16000, Miss.pcmSampleRate(0b1000))
        assertEquals(16000, Miss.pcmSampleRate(0b1110000))
        assertEquals(8000, Miss.pcmSampleRate(0b10000000)) // bit 7 is outside the field
    }

    @Test
    fun `PROTO-11 cameras are devices whose model contains camera`() {
        assertTrue(isCamera("chuangmi.camera.077ac1"))
        assertTrue(isCamera("isa.camera.hlc7"))
        assertEquals(false, isCamera("zhimi.airpurifier.ma2"))
        assertEquals(false, isCamera("yeelink.light.lamp4"))
    }
}
